File: ledgerlm/ledgerlm/server/utils/entraToken.ts
/**
 * Azure Entra ID token fetcher for PostgreSQL Managed Identity auth.
 * Calls the Azure Instance Metadata Service (IMDS) — only available inside
 * Azure App Service / VM. Never called on Replit or local dev.
 *
 * Token is cached in-process and refreshed 5 minutes before expiry.
 * pg.Pool calls getEntraToken() as an async password function, so each
 * new connection automatically gets a fresh token when needed.
 */

interface TokenCache {
  token: string;
  expiresAt: number; // unix milliseconds
}

const REFRESH_BUFFER_MS = 5 * 60 * 1000; // refresh 5 min before expiry
const IMDS_URL =
  'http://169.254.169.254/metadata/identity/oauth2/token' +
  '?api-version=2019-08-01' +
  '&resource=https%3A%2F%2Fossrdbms-aad.database.windows.net';

let tokenCache: TokenCache | null = null;

export async function getEntraToken(): Promise<string> {
  const now = Date.now();

  // Return cached token if still fresh
  if (tokenCache && tokenCache.expiresAt - now > REFRESH_BUFFER_MS) {
    return tokenCache.token;
  }

  console.log('[EntraToken] Fetching new token from Azure IMDS...');

  const res = await fetch(IMDS_URL, {
    headers: { Metadata: 'true' },
    signal: AbortSignal.timeout(10_000), // 10s timeout
  });

  if (!res.ok) {
    const body = await res.text().catch(() => '');
    throw new Error(
      `[EntraToken] IMDS token fetch failed: HTTP ${res.status} — ${body}`
    );
  }

  const data = (await res.json()) as {
    access_token: string;
    expires_on: string; // unix seconds as string
  };

  tokenCache = {
    token: data.access_token,
    expiresAt: parseInt(data.expires_on, 10) * 1000,
  };

  const expiresInMin = Math.round((tokenCache.expiresAt - now) / 60_000);
  console.log(`[EntraToken] Token fetched. Expires in ~${expiresInMin} minutes.`);

  return tokenCache.token;
}

/** For the admin status panel — returns cache state without fetching */
export function getTokenStatus(): {
  cached: boolean;
  expiresInMs?: number;
  expiresInMin?: number;
} {
  if (!tokenCache) return { cached: false };
  const expiresInMs = tokenCache.expiresAt - Date.now();
  return {
    cached: true,
    expiresInMs,
    expiresInMin: Math.round(expiresInMs / 60_000),
  };
}

File: ledgerlm/ledgerlm/python_backend/utils/entra_token.py
"""
Azure Entra ID token fetcher for PostgreSQL Managed Identity auth.
Calls the Azure Instance Metadata Service (IMDS) — only available inside
Azure App Service / VM. Never called on Replit or local dev.

Token is cached in-process and refreshed 5 minutes before expiry.
get_entra_token() is called inside get_db_connection() so every new
psycopg2 connection gets a fresh token when needed.
"""

import time
import logging
import requests

logger = logging.getLogger(__name__)

IMDS_URL = (
    "http://169.254.169.254/metadata/identity/oauth2/token"
    "?api-version=2019-08-01"
    "&resource=https://ossrdbms-aad.database.windows.net"
)
REFRESH_BUFFER_SECONDS = 5 * 60  # refresh 5 min before expiry

_token_cache: dict = {}  # keys: token, expires_at


def get_entra_token() -> str:
    """Fetch (or return cached) Azure Entra access token for Postgres auth."""
    now = time.time()
    expires_at = _token_cache.get("expires_at", 0)

    if _token_cache.get("token") and expires_at - now > REFRESH_BUFFER_SECONDS:
        return _token_cache["token"]

    logger.info("[EntraToken] Fetching new token from Azure IMDS...")

    try:
        resp = requests.get(
            IMDS_URL,
            headers={"Metadata": "true"},
            timeout=10,
        )
        resp.raise_for_status()
    except requests.RequestException as exc:
        raise RuntimeError(
            f"[EntraToken] IMDS token fetch failed: {exc}"
        ) from exc

    data = resp.json()
    _token_cache["token"] = data["access_token"]
    _token_cache["expires_at"] = int(data["expires_on"])  # unix seconds

    expires_in_min = round((_token_cache["expires_at"] - now) / 60)
    logger.info(f"[EntraToken] Token fetched. Expires in ~{expires_in_min} minutes.")

    return _token_cache["token"]


def get_token_status() -> dict:
    """Return cache state without fetching — for health/status endpoints."""
    if not _token_cache.get("token"):
        return {"cached": False}
    expires_in_s = _token_cache["expires_at"] - time.time()
    return {
        "cached": True,
        "expires_in_seconds": round(expires_in_s),
        "expires_in_minutes": round(expires_in_s / 60),
    }
